velox/functions/prestosql/DataSizeFunctions.h

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <map>
+#include <string>
+#include "velox/functions/Macros.h"
+#include "velox/type/DecimalUtil.h"
+
+namespace facebook::velox::functions {
+
+enum class Unit : int128_t {
+  BYTE = 1,
+  KILOBYTE = int128_t(1) << 10, // 2^10
+  MEGABYTE = int128_t(1) << 20, // 2^20
+  GIGABYTE = int128_t(1) << 30, // 2^30
+  TERABYTE = int128_t(1) << 40, // 2^40
+  PETABYTE = int128_t(1) << 50, // 2^50
+  EXABYTE = int128_t(1) << 60, // 2^60
+  ZETTABYTE = (int128_t(1) << 60) * (int128_t(1) << 10), // 2^70
+  YOTTABYTE = (int128_t(1) << 60) * (int128_t(1) << 20) // 2^80
+};
+
+inline Unit parseUnit(const std::string& unitString) {
+  static const std::map<std::string, Unit> unitMap = {
+      {"B", Unit::BYTE},
+      {"kB", Unit::KILOBYTE},
+      {"MB", Unit::MEGABYTE},
+      {"GB", Unit::GIGABYTE},
+      {"TB", Unit::TERABYTE},
+      {"PB", Unit::PETABYTE},
+      {"EB", Unit::EXABYTE},
+      {"ZB", Unit::ZETTABYTE},
+      {"YB", Unit::YOTTABYTE}};
+  auto it = unitMap.find(unitString);
+  VELOX_USER_CHECK(it != unitMap.end(), "Invalid data size: '{}'", unitString);
+  return it->second;
+}
+
+inline double parseValue(const std::string& value) {
+  try {
+    return std::stod(value);
```

Comment:
presto java uses BigDecimal to convert this value, using double here can result in double precision issues.

An equivalent solution here would meant that we pull out the cast ([code pointer](https://github.com/facebookincubator/velox/blob/c6cbeeeaf0f708904b3e693b1a33d36299f6cf56/velox/expression/CastExpr-inl.h#L96)) and arithmetic ([code pointer](https://github.com/facebookincubator/velox/blob/c6cbeeeaf0f708904b3e693b1a33d36299f6cf56/velox/functions/sparksql/DecimalArithmetic.cpp#L332)) implementations for Decimal type into utility functions and them here directly.

@kgpai @Yuhta Any thoughts on whether its worth the effort ?

Reply:
@natashasehgal whatever way gets decided here, the output would either be in double or decimal types(initially converted from string) and in either case it would be good to take a look at the cast implementation to make sure type conversion is accurate. Code pointer: https://github.com/facebookincubator/velox/blob/c6cbeeeaf0f708904b3e693b1a33d36299f6cf56/velox/expression/CastExpr-inl.h#L629

Reply:
@natashasehgal I synced up with @kgpai offline and he agrees we need to convert the input to a decimal and use decimal arithmetic when scaling it up to the specified Unit similar to what presto java does.

Can you take a look at the code pointers i mentioned in the previous comments for how you can convert and add decimal types?

Reply:
I copied this code. One issue is `toDecimalValue(StringView(str), precision, scale, result)` expects result to be of type int128_t (this results in a loss of precision). There's an error thrown if I try to set result to DecimalType/Long Decimal due from [line](https://github.com/facebookincubator/velox/blob/c6cbeeeaf0f708904b3e693b1a33d36299f6cf56/velox/expression/CastExpr-inl.h#L185)

Any thoughts about converting value to int128_t (e.g. 2.2 -> 22), and rounding down while multiplying? Is overflow a concern here?

cc @karteekmurthys @bikramSingh91